flink-runtime/src/main/java/org/apache/flink/runtime/security/FlinkSecurityManager.java

```diff
@@ -0,0 +1,220 @@
+                                ClusterOptions.INTERCEPT_USER_SYSTEM_EXIT.defaultValue(),
+                                ClusterOptions.HALT_ON_SYSTEM_EXIT.key(),
+                                ClusterOptions.HALT_ON_SYSTEM_EXIT.defaultValue(),
+                                e));
+            }
+        }
+        FlinkSecurityManager.flinkSecurityManager = flinkSecurityManager;
+    }
+
+    public static void monitorUserSystemExitForCurrentThread() {
+        if (FlinkSecurityManager.flinkSecurityManager != null) {
+            FlinkSecurityManager.flinkSecurityManager.monitorUserSystemExit();
+        }
+    }
+
+    public static void unmonitorUserSystemExitForCurrentThread() {
+        if (FlinkSecurityManager.flinkSecurityManager != null) {
+            FlinkSecurityManager.flinkSecurityManager.unmonitorUserSystemExit();
+        }
+    }
+
+    @Override
+    public void checkPermission(Permission perm) {
+        if (originalSecurityManager != null) {
+            originalSecurityManager.checkPermission(perm);
+        }
+    }
+
+    @Override
+    public void checkPermission(Permission perm, Object context) {
+        if (originalSecurityManager != null) {
+            originalSecurityManager.checkPermission(perm, context);
+        }
+    }
+
+    @Override
+    public void checkExit(int status) {
+        if (userSystemExitMonitored()) {
+            switch (userSystemExitMode) {
+                case DISABLED:
+                    break;
+                case LOG:
+                    // Add exception trace log to help users to debug where exit came from.
+                    LOG.warn(
+                            "Exiting JVM with status {} is monitored, logging and exiting",
+                            status,
+                            new UserSystemExitException());
+                    break;
+                case THROW:
+                    throw new UserSystemExitException();
+                default:
+                    // Must not happen if exhaustively handling all modes above. Logging as being
+                    // already at exit path.
+                    LOG.warn("No valid check exit mode configured: {}", userSystemExitMode);
+            }
+        }
+        // As this security manager is current at outer most of the chain and it has exit guard
+        // option, invoke inner security manager here after passing guard checking above, if any.
+        if (originalSecurityManager != null) {
+            originalSecurityManager.checkExit(status);
+        }
+        // At this point, exit is determined. Invoke exit handler if defined, otherwise check ended.
+        if (onExitBehavior != null) {
+            // Unset ourselves to allow exiting.
+            System.setSecurityManager(null);
+            // Execute the desired behavior, e.g. forceful exit instead of proceeding with
+            // System.exit().
+            onExitBehavior.accept(status);
```

Comment:
Why can't we simplify this to
```java
if (haltOnSystemExit) {
  System.setSecurityManager(null);
  Runtime.getRuntime().halt(status);
}
```

Reply:
I was wondering the same as the halt is the only functionality, we use so far. It would make testing harder, though.

Reply:
I had the same decision point, but I rather preserved the existing implementation for the existing tests as well.

Reply:
Okay, understood